### PR TITLE
added go lint

### PR DESCRIPTION
**Pull request by shoebsd31**
Hi, 

With regards to #31 
I have added go lint to support multiple OS 

Currently it has some linting errors. I tried to resolve them but it was breaking. Can you please help me with resolving the linting errors

### CODE REVIEW

**Review by connorkuehl — CHANGES_REQUESTED** [review]
Hi @shoebsd31! Thank you for setting this up!

I have a couple of comments:

I appreciate the testing matrix approach you took, but I don't think we
need anything as advanced as a testing matrix across multiple operating
systems. I would rather just have it run against a Linux distribution.

So, I would like to merge this in a more simpler form. I think perhaps
something like this would work (I haven't tested it):

```diff
diff --git a/.github/workflows/go.yml b/.github/workflows/go.yml
index 1bd0a1f..e5b0b8f 100644
--- a/.github/workflows/go.yml
+++ b/.github/workflows/go.yml
@@ -23,3 +23,12 @@ jobs:

     - name: Test
       run: go test -v ./...
+
+  lint:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: golangci-lint
+      uses: golangci/golangci-lint-action@v2
+      with:
+        version: v1.29
```

Also, here's some advice on the linting errors:

```diff
diff --git a/command.go b/command.go
index 424abbe..a66b140 100644
--- a/command.go
+++ b/command.go
@@ -81,7 +81,7 @@ func CheckKarma(req poppleRequest, rsp poppleResponse, db *gorm.DB) {

        reply := strings.Builder{}

-       for subject, _ := range subjects {
+       for subject := range subjects {
                var entity Entity
                db.Where(&Entity{GuildID: req.GuildID(), Name: subject}).First(&entity)
                reply.WriteString(fmt.Sprintf("%s%s has %d karma.", sep, subject, entity.Karma))
```

For the first one, it looks like the `_` underscore is unnecessary for
omitting the value from the (key, value) pair when iterating over a map.

Seems like the underscore can be left out and only have one item on the
left side of the assignment statement to the `range` keyword and it will
just iterate over the keys of the map, which is what we want here.

```diff
diff --git a/parse.go b/parse.go
index d95e03a..37bdba5 100644
--- a/parse.go
+++ b/parse.go
@@ -189,8 +189,7 @@ type item struct {
 type itemType int
 
 const (
-       itemError        itemType = iota // error occurred during lexing
-       itemText                         // alphanumeric
+       itemText         itemType = iota // alphanumeric
        itemTextInParens                 // (alphanumeric)
 )
```

For this one, deleting `itemError` also removes the type information
from these "enum-like" constants. To fix this, the new item at the
top of the list must include the type information.

```diff
diff --git a/popple.go b/popple.go
index 664b23f..18d30e2 100644
--- a/popple.go
+++ b/popple.go
@@ -63,8 +63,8 @@ func main() {
                log.Fatalf("Failed to read token from %s\n", *tokenFile)
        }

-       db.AutoMigrate(&Config{})
-       db.AutoMigrate(&Entity{})
+       _ = db.AutoMigrate(&Config{})
+       _ = db.AutoMigrate(&Entity{})

        session, err := discordgo.New("Bot " + string(token))
        if err != nil {
```

In Go, you only use the `:=` operator for assignment when there's a new
variable on the lefthand side of the assignment operator. In this case,
I guess Go considers the discard underscore (`_`) to not be a new
variable, so you can simply use `=` instead of `:=`.

I would also prefer if your PR did one of these two suggestions:

* The CI job and the lint fixes are in the same commit

or

* The lint fixes are in one commit and the CI job is added in the
  commit after the lint fixes.

Hope this helps! Thanks for taking this on. Let me know if I can be of
any more help. 🙂

**Comment by shoebsd31** [comment]
> Hi @shoebsd31! Thank you for setting this up!
> 
> I have a couple of comments:
> 
> I appreciate the testing matrix approach you took, but I don't think we
> need anything as advanced as a testing matrix across multiple operating
> systems. I would rather just have it run against a Linux distribution.
> 
> So, I would like to merge this in a more simpler form. I think perhaps
> something like this would work (I haven't tested it):
> 
> ```diff
> diff --git a/.github/workflows/go.yml b/.github/workflows/go.yml
> index 1bd0a1f..e5b0b8f 100644
> --- a/.github/workflows/go.yml
> +++ b/.github/workflows/go.yml
> @@ -23,3 +23,12 @@ jobs:
> 
>      - name: Test
>        run: go test -v ./...
> +
> +  lint:
> +    runs-on: ubuntu-latest
> +    steps:
> +    - uses: actions/checkout@v2
> +    - name: golangci-lint
> +      uses: golangci/golangci-lint-action@v2
> +      with:
> +        version: v1.29
> ```
> 
> Also, here's some advice on the linting errors:
> 
> ```diff
> diff --git a/command.go b/command.go
> index 424abbe..a66b140 100644
> --- a/command.go
> +++ b/command.go
> @@ -81,7 +81,7 @@ func CheckKarma(req poppleRequest, rsp poppleResponse, db *gorm.DB) {
> 
>         reply := strings.Builder{}
> 
> -       for subject, _ := range subjects {
> +       for subject := range subjects {
>                 var entity Entity
>                 db.Where(&Entity{GuildID: req.GuildID(), Name: subject}).First(&entity)
>                 reply.WriteString(fmt.Sprintf("%s%s has %d karma.", sep, subject, entity.Karma))
> ```
> 
> For the first one, it looks like the `_` underscore is unnecessary for
> omitting the value from the (key, value) pair when iterating over a map.
> 
> Seems like the underscore can be left out and only have one item on the
> left side of the assignment statement to the `range` keyword and it will
> just iterate over the keys of the map, which is what we want here.
> 
> ```diff
> diff --git a/parse.go b/parse.go
> index d95e03a..37bdba5 100644
> --- a/parse.go
> +++ b/parse.go
> @@ -189,8 +189,7 @@ type item struct {
>  type itemType int
>  
>  const (
> -       itemError        itemType = iota // error occurred during lexing
> -       itemText                         // alphanumeric
> +       itemText         itemType = iota // alphanumeric
>         itemTextInParens                 // (alphanumeric)
>  )
> ```
> 
> For this one, deleting `itemError` also removes the type information
> from these "enum-like" constants. To fix this, the new item at the
> top of the list must include the type information.
> 
> ```diff
> diff --git a/popple.go b/popple.go
> index 664b23f..18d30e2 100644
> --- a/popple.go
> +++ b/popple.go
> @@ -63,8 +63,8 @@ func main() {
>                 log.Fatalf("Failed to read token from %s\n", *tokenFile)
>         }
> 
> -       db.AutoMigrate(&Config{})
> -       db.AutoMigrate(&Entity{})
> +       _ = db.AutoMigrate(&Config{})
> +       _ = db.AutoMigrate(&Entity{})
> 
>         session, err := discordgo.New("Bot " + string(token))
>         if err != nil {
> ```
> 
> In Go, you only use the `:=` operator for assignment when there's a new
> variable on the lefthand side of the assignment operator. In this case,
> I guess Go considers the discard underscore (`_`) to not be a new
> variable, so you can simply use `=` instead of `:=`.
> 
> I would also prefer if your PR did one of these two suggestions:
> 
> * The CI job and the lint fixes are in the same commit
> 
> or
> 
> * The lint fixes are in one commit and the CI job is added in the
>   commit after the lint fixes.
> 
> Hope this helps! Thanks for taking this on. Let me know if I can be of
> any more help. 🙂

thanks for your comment. I will update the code in the weekend :)

**Comment by shoebsd31** [comment]
Hi @connorkuehl , can you please review the changes

File: parse.go [hunk 1]
```diff
@@ -188,9 +188,9 @@ type item struct {
 
 type itemType int
 
+//nolint:golint,unused
```
Comment (connorkuehl) [inline parse.go:191]: Please remove this comment.

**Review by connorkuehl — CHANGES_REQUESTED** [review]
Hi @shoebsd31! Thanks! It's looking a lot better. I just have 1 comment inline and then otherwise, please rebase onto the latest version of my `master` branch and squash all of your commits into 1 commit and then I will merge your PR.

**Comment by shoebsd31** [comment]
> Hi @shoebsd31! Thanks! It's looking a lot better. I just have 1 comment inline and then otherwise, please rebase onto the latest version of my `master` branch and squash all of your commits into 1 commit and then I will merge your PR.

Or else I will just create a new PR, you can discard the current one

**Comment by connorkuehl** [comment]
> > Hi @shoebsd31! Thanks! It's looking a lot better. I just have 1 comment inline and then otherwise, please rebase onto the latest version of my `master` branch and squash all of your commits into 1 commit and then I will merge your PR.
> 
> Or else I will just create a new PR, you can discard the current one

Sure, I don't mind either way if you want to force push to to this branch or open a new PR 🙂

**Comment by connorkuehl** [comment]
Closing per comment in #40.

**Pull request closed without merging**